SCons/Util/stats.py

```diff
@@ -0,0 +1,212 @@
+
+    def do_print(self):
+        raise NotImplementedError
+
+    def enable(self, outfp):
+        self.outfp = outfp
+        self.append = self.do_append
+        self.print_stats = self.do_print
+        self.enabled = True
+
+    def do_nothing(self, *args, **kw):
+        pass
+
+
+class CountStats(Stats):
+
+    def __init__(self):
+        super().__init__()
+        self.stats_table = {}
+
+    def do_append(self, label):
+        self.labels.append(label)
+        self.stats.append(SCons.Debug.fetchLoggedInstances())
+
+    def do_print(self):
+        self.stats_table = {}
+        for s in self.stats:
+            for n in [t[0] for t in s]:
+                self.stats_table[n] = [0, 0, 0, 0]
+        i = 0
+        for s in self.stats:
+            for n, c in s:
+                self.stats_table[n][i] = c
+            i = i + 1
+        self.outfp.write("Object counts:\n")
+        pre = ["   "]
+        post = ["   %s\n"]
+        l = len(self.stats)
+        fmt1 = ''.join(pre + [' %7s'] * l + post)
+        fmt2 = ''.join(pre + [' %7d'] * l + post)
+        labels = self.labels[:l]
+        labels.append(("", "Class"))
+        self.outfp.write(fmt1 % tuple([x[0] for x in labels]))
+        self.outfp.write(fmt1 % tuple([x[1] for x in labels]))
+        for k in sorted(self.stats_table.keys()):
+            r = self.stats_table[k][:l] + [k]
+            self.outfp.write(fmt2 % tuple(r))
+
+
+class MemStats(Stats):
+    def do_append(self, label):
+        self.labels.append(label)
+        self.stats.append(SCons.Debug.memory())
+
+    def do_print(self):
+        fmt = 'Memory %-32s %12d\n'
+        for label, stats in zip(self.labels, self.stats):
+            self.outfp.write(fmt % (label, stats))
+
+
+class TimeStats(Stats):
+    def __init__(self):
+        super().__init__()
+        self.totals = {}
+        self.commands = {}  # we get order from insertion order, and can address individual via dict
+
+    def total_times(self, build_time, sconscript_time, scons_exec_time, command_exec_time):
+        self.totals = {
+            'build_time': build_time,
+            'sconscript_time': sconscript_time,
+            'scons_exec_time': scons_exec_time,
+            'command_exec_time': command_exec_time
+        }
+
+    def add_command(self, command, start_time, finish_time):
+        if command in self.commands:
+            print("Duplicate command %s" % command)
+        self.commands[command] = {'start': start_time,
+                                  'end' : finish_time,
+                                  'duration': finish_time - start_time}
+
+
+COUNT_STATS = CountStats()
+MEMORY_STATS = MemStats()
+TIME_STATS = TimeStats()
+
+
+
+def WriteJsonFile():
+    """
+    Actually write the JSON file with debug information.
+    Depending which of : count, time, action-timestamps,memory their information will be written.
+    """
+
+    from SCons.Script import BUILD_TARGETS, COMMAND_LINE_TARGETS, ARGUMENTS, ARGLIST
```

Comment:
perhaps annotate to avoid checker `import-outside-toplevel` type errors?

Reply:
done.